doc/HOW_THINGS_WORK.md

```diff
@@ -0,0 +1,225 @@
+The goal of this document is to provide high-level overviews of how
+various systems work inside Chef Server. If you encounter a process
+or feature that isn’t straightforward or took you time to understand,
+consider writing an explanation here.
+
+## Search Indexing
+
+Chef Server performs searches by querying a search index. This index
+is updated on every write to the Chef Server. That is, every time we
+write to oc_erchef's postgresql database, we usually also need to
+write to the search index.
+
+This section describes how data gets from erchef to the search index.
+For how we query this data later, see Search Queries.
+
+Chef Server supports two backends for the search index:
+
+- Solr, and
+- Elasticsearch
+
+Solr is the default and is shipped inside the Chef Server package.. It
+has been supported for longer and is more battle tested. ElasticSearch
+is the default for Chef Backend because of its easy-to-use replication
+clustering.
+
+Chef Server has 3 different "search_queue_mode"s:
+
+- rabbitmq (Solr-only)
+- batch (Solr or ES)
+- inline (Solr or ES)
+
+This configurable controls how the chef_index module sends requests to
+the search index. The default is `rabbitmq`.
+
+### Rabbitmq
+
+```
+  +--------+    +----------+    +------------------+    +------+
+  | erchef | -> | rabbitmq | <- | opscode-expander | -> | Solr |
+  +--------+    +----------+    +------------------+    +------+
+```
+
+In the rabbitmq search_queue_mode, erchef places the object to be
+indexed on a rabbitmq queue. It them moves forward with the request
+and returns a response to the user.  Another service named
+`opscode-expander` reads the object from the queue, expands it (see
+Document Expansion for details), and posts it to solr. Note that a
+failure to write to the search index will not trigger a failure of the
+API request that wrote the data since it happens asynchronously via a queue.
```

Comment:
Worth mentioning that a failure to write to the *queue* will fail the request (500) without updating any data.